[Darius Ng] iP

dfordarius opened the pull request:
## DukeChatBot

> "We do not learn from experience. We learn from reflecting on experience" - John Dewey

DukeChatBot frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn 
* ~FAST~ *SUPER FAST* to use

All you need to do is,
1. download it from [here](https://github.com/dfordarius/ip/releases)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And its **FREE**

Features:

- [x] Managing Tasks
- [x] Managing Deadlines
- [x] Reminders 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method: 
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

dfordarius committed:
Implemented Level 1. Greet, Echo, Exit

dfordarius committed:
Implemented Add, List

dfordarius committed:
Implemented Level 3. Mark as Done

dfordarius committed:
Implemented Level 4. ToDos, Events, Deadlines

dfordarius committed:
Added Automated Text UI Testing

dfordarius committed:
Implemented Level-5 Handle Errors

dfordarius committed:
Implemented Level 6 Delete

dfordarius committed:
Implemented the use of Enums

dfordarius committed:
Implement Level-7 Save Features

Duke is runs on data that is currently not saved and resets after each run.

Saves data into user's root directory/data/data.txt

dfordarius committed:
Implement Level-8: Date and Time

Duke could not handle date formats

add parser to properly format dates and time for classes such as deadline and event

dfordarius committed:
Implement Level 8: Date and Time

Duke could not handle date and time formats

Add a parser to properly format date and times for classes such as deadline and events

dfordarius committed:
Merge branch 'branch-Level-7'

dfordarius committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

dfordarius committed:
Add A-MoreOOP: Use More OOP

dfordarius committed:
Move classes into respective packages

Classes were not in packages

Code looks messy and tidy without classifying them into packages

Lets move classes into packages

dfordarius committed:
Merge remote-tracking branch 'origin/add-gradle-support'

dfordarius committed:
Add Gradle support

dfordarius committed:
Add JUnit Test

Did not have any JUnit to test the behaviour of the code

Add JUnit testing for Classes such as Deadline, Event and Todo

dfordarius committed:
Add javadocs

Duke does not have javadocs and it would be hard for users to understand its codes.

Add javadocs so that its code is easier to understand and read.

dfordarius committed:
Merge remote-tracking branch 'origin/branch-Level-8' into branch-Level-8

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

dfordarius committed:
Added A-CodingStandard

dfordarius committed:
Add Level-9

dfordarius committed:
Merge branch 'branch-A-JavaDoc'

dfordarius committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

dfordarius committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java
#	src/main/java/duke/tasks/Task.java

dfordarius committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/duke/Storage.java

dfordarius committed:
Updated JavaDocs

dfordarius committed:
Updated JavaDocs and cleaned code

dfordarius committed:
Add Checkstyle

Currently no checkstyle in the program.

Checkstyle aids us to follow the coding standard.

Add checkstyle.xml and suppreession.xml and use gradle for checkstyle.

dfordarius committed:
Fix Checkstyle Errors

dfordarius committed:
Updated javadocs

dfordarius committed:
removed extra line

dfordarius committed:
Fix checkstyle

dfordarius committed:
Add Level-10 GUI

Duke does not have any GUI implemented.

Without GUI Duke looks very plain.

Lets update Duke by adding GUI using JavaFx

dfordarius committed:
Add A-Varargs

Duke does not use any Varargs.

Lets add some Varargs to Duke.

dfordarius committed:
Merge branch 'branch-A-CheckStyle'

dfordarius committed:
Fix Checkstyle

dfordarius committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ListCommand.java

dfordarius committed:
Merge branch 'branch-A-Varargs'

# Conflicts:
#	src/main/java/duke/Ui.java

dfordarius committed:
Fix Checkstyle

dfordarius committed:
Update find implementation

LeonardYam commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,61 @@
+            throw new DukeException("\tYou currently have no tasks in your list to delete.");
+        } else {
+            try {
+                if(isFind) {
```
Should there be a whitespace after the `if` keyword?

dfordarius replied:
Hey leonard, thanks for the review! Will make these changes.

LeonardYam commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,180 @@
+ */
+public class TaskList {
+    private ArrayList<Task> listOfTasks;
+    private static ArrayList<Task> listOfFindTasks;
```
Is the static modifier intended for this variable?

LeonardYam commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+ * Find is a command that finds tasks in the task list filtered by a specific keyword.
+ */
+public class FindCommand extends Command {
+    private String [] taskName;
```
Should the array specifier be attached to the type? (e.g. String[])

LeonardYam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+     * @return Command that corresponds to the user input.
+     * @throws DukeException If the user input cannot be recognised.
+     */
+    public static Command parse(String inputCommand) throws DukeException {
```
Could this be refactored to reduce the method length?

```
// One example
case unmark:
    handleUnmark(splitString)
```

LeonardYam commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+        dialogContainer.getChildren().add(DialogBox.getDukeDialog(Ui.welcomeMessage(), dukeImage));
```
Should this be refactored to simpler expressions by evaluating in steps?

LeonardYam reviewed (commented):
Overall, I felt the coding standards were met - especially on naming methods

dfordarius committed:
Update implementation of Find

dfordarius committed:
Add Assertions to Duke

There are no assertions in Duke

Assertions are used to codify the requirements that render a program correct or not by testing conditions.

Lets add an assertion for the method getResponse.

This helps to ensure the parameter of getResponse is not empty.

dfordarius committed:
Improve CodeQuality

Specifically improved javadocs descriptions

dfordarius committed:
Add Streams

No use of streams in duke.

Streams help to improve readability and condense the code.

Lets use a sequence of stream operations to print the list of tasks for the user.

dfordarius committed:
Add Streams to Duke

No uses of streams in Duke.

Stream helps to condense code and improve readability.

Lets use a series of stream operations to print the list of tasks for the users of Duke.

dfordarius committed:
Merge pull request #1 from dfordarius/A-Assertions

Add Assertions to Duke

dfordarius committed:
Merge pull request #2 from dfordarius/branch-A-CodeQuality

Branch a code quality

dfordarius committed:
Merge pull request #3 from dfordarius/branch-A-Streams

Add Streams to Duke

dfordarius committed:
Create gradle.yml

dfordarius committed:
Update gradle.yml

Removed IO lines which causes problems.

dfordarius committed:
Fixed Checkstyle

dfordarius committed:
Reformatted gradle.yml

dfordarius committed:
edit gradle.yml

dfordarius committed:
Add B-Reminders

Duke currently does not support reminder functions.

Adding reminder functions will allow duke to remind users of upcoming deadlines or event start dates.

Let's add a new command ReminderCommand to allow user to see list of tasks that has upcoming deadlines.

dfordarius committed:
Merge branch 'branch-B-Reminders'

ongweijie7 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+    private final String filePath;
+
+    /**
+     * Constructor for Storage.
+     *
+     * @param filePath The filepath where data is saved.
+     */
```
I noticed you did a good job at ensuring every fullstops at the end of each statement for javadocs which might have been commonly overlooked 👍 👍

ongweijie7 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+ */
+public class TaskList {
+    private ArrayList<Task> listOfTasks;
+    private ArrayList<Task> listOfFindTasks;
```
Maybe this could have been
```suggestion
    private ArrayList<Task> listOfFoundTasks;
```
Since 'find' is a verb and using an adjective can make it less confusing

ongweijie7 commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class DeadlineCommand extends Command {
+    private String description;
+    private String by;
```
Could this have been better named to avoid confusion?

ongweijie7 reviewed (commented):
In general,  I noticed that you named most of your classes and variables in a rational manner which made it easy to follow and understand. The naming also abided mostly with coding standards. Good Job!! 💯 💯

dfordarius committed:
Improve Code Quality according to bot.

There were some areas where code quality could have been much better.

Lets improve the code quality.

dfordarius committed:
Change personality of Duke

Duke has a very static personality.

Lets add more expressions to duke's replies.

dfordarius committed:
Add User guide, Ui

Duke does not have a user guide.

Lets add a user guide so that users can get familiar with duke's features.

dfordarius committed:
Add more personality to duke replies.

Currently duke's expression in his replies are very monotonic.

Lets add more expressions to dukes replies, so he sounds friendlier

dfordarius committed:
Update Ui.png and clean code.

Duke's UI is currently very plain.

Lets improve the UI and clean the code up.

dfordarius committed:
Update README.md

dfordarius committed:
Update README.md
